docs/en/sql-reference/functions/array-functions.md

```diff
@@ -2118,6 +2118,113 @@ Result:
 └─────────────────────┘
 ```
 
+
+## randomSampleFromArray
+
+Returns a random sample from an input array. The number of elements in the sample is determined by the `sampleSize` argument. The function ensures that each element in the input array has an equal chance of being included in the sample.
+
+**Syntax**
+
+```sql
+randomSampleFromArray(arr, sampleSize)
+```
+
+**Arguments**
+
+- `arr` — The input array from which to sample elements, including nested arrays.
+- `sampleSize` — The number of elements to include in the random sample.
+
+**Returned value**
+
+- An array containing a random sample of elements from the input array.
+
+**Examples**
+
+Query:
+
+```sql
+SELECT randomSampleFromArray([1, 2, 3, 4, 5, 6], 3) as res;
+```
+
+Result:
+```
+┌─res─────┐
+│ [2,3,1] │
+└─────────┘
+or
+┌─res─────┐
+│ [5,2,6] │
+└─────────┘
+```
+
+Query:
+
+```sql
+SELECT randomSampleFromArray(['apple', 'banana', 'cherry', 'date'], 2) as res;
+```
+
+Result:
+```
+┌─res────────────────┐
+│ ['apple','banana'] │
+└────────────────────┘
+or
+┌─res────────────────┐
+│ ['banana','apple'] │
+└────────────────────┘
+or
+┌─res───────────────┐
+│ ['cherry','date'] │
+└───────────────────┘
+```
+
+Query:
+
+```sql
+SELECT randomSampleFromArray([[1, 2], [3, 4], [5, 6]], 2) as res;
+```
+
+Result:
+```
+┌─res───────────┐
+│ [[1,2],[5,6]] │
+└───────────────┘
+or
+┌─res───────────┐
+│ [[3,4],[5,6]] │
+└───────────────┘
+```
+
+Query:
+
+```sql
+SELECT randomSampleFromArray([1, 2, 3, 4, 5], 0) as res;
+```
+
+Result:
+```
+TODO: FIX ME
```

Comment:
Return an empty array?

Reply:
Yes, but when I am running this query:

`clickhouse-client --query "SELECT randomSampleFromArray([[1,2,3], [3,4], [6,9]], 0);`

I am getting this error:

```
Received exception from server (version 23.9.1):
Code: 9. DB::Exception: Received from localhost:9000. DB::Exception: Incorrect size of nested column in constructor of ColumnConst: 0, must be 1.: While processing randomSampleFromArray([[1, 2, 3], [3, 4], [6, 9]], 0). (SIZES_OF_COLUMNS_DOESNT_MATCH)
(query: SELECT randomSampleFromArray([[1,2,3], [3,4], [6,9]], 0);
)
```

I will investigate it.

Reply:
I tried it, but the exception did not come from the execute method.

Reply:
Function arguments can generally be const or non-const. For example, a const string and a non-const string have different internal representations. With one function parameter, that's 2 variants to implement, with two function parameters, it's 4, with three function parameters, it's 8 etc. Usually, not all combinations need to be implemented as certain function parameters are usually only const (in this case, the sampleCount parameter is always const).

If `useDefaultImplementationForConstants = true`, then ClickHouse converts an invocation with only const parameters into an invocation with only non-const parameters. This basically happens with `arrayRandomSample(['a', 'b', 'c'], 2)`. The motivation for that is to reduce the number of variants that the programmer needs to implement. The problem is that something goes wrong during the conversion here (didn't check in detail).

We can look at the way function `arrayRotateLeft` is implemented. Turns out that they set  `useDefaultImplementationForConstants = false` (i.e. inherit the default by not overriding the virtual method) and implement the const/non-const separately. Maybe that's an option for you too.

Reply:
Thanks for the detailed answer :slightly_smiling_face: . It was helpful information for me. 
I will take a look.

Reply:
In the end, this line solved the problem:
```
if (K == 0)
  {
      // Handle edge cases where input array is empty or K is 0
      res_data->getOffsets().push_back(0);
      return res_data;
  }
```

Reply:
Sounds good. Let me quickly verify locally, we can merge then.

(the optional seed parameter is perhaps overkill - it will only be useful for testing but your tests cater for randomness quite well.)

Reply:
OK :) 
I still need to iterate on all your open comments and solve them.
And one more case that I need to check:
Currently, my code does not support multi- rows implementation. 
I will solve it (including tests), and then I will ping you.

Reply:
Unrelated comment: Could you please chmod +x to the `.sh` test file. Otherwise, `clickhouse-test` ([docs](https://clickhouse.com/docs/en/development/tests#functional-test-locally)) complains for me about permissions.

Reply:
It also looks like you would need to re-generate the expected result file. It is out-of-sync with the test.

Reply:
What doesn't work yet are
- `select arrayRandomSample(['a', 'b', 'c'], materialize(1))` and
- `select arrayRandomSample(materialize(['a', 'b', 'c']), 1)`

Reply:
Yes, exactly, I saw it. 
2 minutes. 
Thanks!

Reply:
I have not succeeded in treating all comments yet. One of them is the "materialize" comment. 
Now, I will take a look.

Reply:
The empty array problem - solved
The materialize function is blocking until we fix it with flow-up PR.

Reply:
Unfortuantely, the materialize problem blocks merge. It is relevant for 99% of all practical use cases, basically whenever the function is applied to a column of arrays. Not supporting it properly raise eyebrows with many users.

But no hurry to finish this PR up. Either you can do it in 2, 3, ... weeks, or if not let me know, I can also finish it (when I find some time).

Reply:
Yes, you are right; I will fix it with this PR.